### ISSUE TITLE
CSS Custom Properties (var(--foo)) in theme secondary colour causes fatal error

### ISSUE THREAD

**Issue by bennypowers** [storybookjs/storybook#6135]
**Describe the bug**
Using CSS variables in theme breaks storybook.

**To Reproduce**
Use this theme:

```js
import { create } from '@storybook/theming';

export default create({
  base: 'light',

  colorSecondary: 'var(--my-var, red)',

})
```
**Expected behavior**
I expect the styles to be applied.

**Code snippets**
```
The above error occurred in the <Context.Consumer> component:
    in Styled(a) (created by Link)
    in Link (created by KnobPanel)
    in div (created by Context.Consumer)
    in Styled(div) (created by Placeholder)
    in div (created by Context.Consumer)
    in Styled(div) (created by Placeholder)
    in Placeholder (created by KnobPanel)
    in KnobPanel
    in div (created by Context.Consumer)
    in Styled(div)
    in div (created by Context.Consumer)
    in Styled(div)
    in Unknown
    in Unknown
    in Unknown
    in Unknown (created by Context.Consumer)
    in _default (created by Layout)
    in div (created by Context.Consumer)
    in Styled(div) (created by Panel)
    in Panel (created by Layout)
    in div (created by Context.Consumer)
    in Styled(div) (created by Main)
    in div (created by Context.Consumer)
    in Styled(div) (created by Main)
    in Main (created by Layout)
    in Layout (created by Context.Consumer)
    in WithTheme(Layout)
    in Unknown
    in Unknown (created by ResizeDetector)
    in ChildWrapper (created by ResizeDetector)
    in ResizeDetector
    in div (created by Context.Consumer)
    in Styled(div)
    in Unknown
    in Unknown (created by Manager)
    in ThemeProvider (created by Manager)
    in Manager (created by Context.Consumer)
    in Location (created by QueryLocation)
    in QueryLocation (created by Root)
    in LocationProvider (created by Root)
    in HelmetProvider (created by Root)
    in Root

React will try to recreate this component tree from scratch using the error boundary you provided, LocationProvider. vendors~main.5bfe49427a0ecbcae866.bundle.js:155228:5
The above error occurred in the <LocationProvider> component:
    in LocationProvider (created by Root)
    in HelmetProvider (created by Root)
    in Root

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries. vendors~main.5bfe49427a0ecbcae866.bundle.js:155228:5
Error: Couldn't parse the color string. Please provide the color as a string in hex, rgb, rgba, hsl or hsla notation.

vendors~main.5bfe49427a0ecbcae866.bundle.js:19244:14
Error: Couldn't parse the color string. Please provide the color as a string in hex, rgb, rgba, hsl or hsla notation.
```

**System:**
 - OS: MacOS
 - Device: Macbook Pro 2018
 - Browser: Firefox, Chrome
 - Framework: Web Components
 - Addons: theming
 - Version: 5.1.0-alpha.7

**Additional context**
You can try this yourself just by setting the secondary color prop in react dev tools under ThemeProvider

**Comment by Armanio** [comment]
Storybook under hood use [polished.js](https://polished.js.org/) for transforming theme colors: eg set opacity, darken color, etc. 
Unfortunately, now polished not support css vars. 
I see 2 ways for solve this problem: 
- don't use css vars in storybook (probably we set it in doc 🤔)
- not use polished and transforming css manually (it big question...)

**Comment by bennypowers** [comment]
CSS Custom Properties have been supported in browsers since 2014. It seems less than ideal to commit "Our CSS-in-JS library doesn't support browser standards" to documentation.

**Comment by Armanio** [comment]
Of course you are right. I will think about it. Maybe it is not as difficult as it seems.

**Comment by bennypowers** [comment]
If I can help, please let me know.

**Comment by Armanio** [comment]
Any PR is welcome. You can try to support CSS variables yourself. 🥇 
Anywhere tomorrow i investigate this question.

**Comment by bennypowers** [comment]
@Armanio did your investigations turn up anything helpful?

**Comment by Armanio** [comment]
@bennypowers honestly, no. Sorry. 😥 
But first, I'll want to get @shilman and @ndelangen opinions.
SB use 'polished' not many times:
<img width="435" alt="Снимок экрана 2019-04-14 в 1 46 25" src="https://user-images.githubusercontent.com/3195714/56086065-71726980-5e57-11e9-885e-760eac036aac.png">

Some features can realize by native CSS (with rgba, filters, etc). But not all. It's can be a design problem and @domyen will be unhappy. 
In conclusion, I think it's a good point for SB future:
- it reduces bundle size 
- CSS vars has a good browser support (>90%)
- change theme colors in runtime

Anyway, wait for maintainers thoughts.

**Comment by shilman** [comment]
Thanks for investigating @Armanio.  Since it's not used much we should be able to switch it out  for something more permissive without too much hassle. But I'll defer to @ndelangen and @domyen on theming decisions since they did the work, and will be maintaining our themes

**Comment by bennypowers** [comment]
I define them in a css file that's linked in `preview-head.html` 

The good-old-fashioned way :D

**Comment by Hypnosphi** [comment]
On root level?

**Comment by bennypowers** [comment]
apologies, it's manager-head, not preview-head. this is my `manager-head.html`

```html
<link rel="stylesheet" href="./my-theme.css"></link>
```

and here's my-theme.css:

```css

html,
body {
  padding: 0;
  margin: 0;
  font-family: var(--ftr-font-family-primary);
}

html {
  --ftr-font-family-primary: "proxima-nova", arial, sans-serif;

  /*  General */
  --white: #fff;
  --white-rgb: 255, 255, 255;
  --black: #000;
  --black-rgb: 0, 0, 0;
/* etc*/
}
```

**Comment by Hypnosphi** [comment]
OK, if you move it to `preview-head.html`, you can work around your issue like this.

```
import { create } from '@storybook/theming';

const var = (property, default) =>
  getComputedStyle(document.documentElement).getPropertyValue(property) || default

export default create({
  base: 'light',

  colorSecondary: var('--my-var', 'red'),

})
```

**Comment by bennypowers** [comment]
In my case, though, I want to theme the storybook manager. I don't need to dynamically change values later on.

**Comment by Hypnosphi** [comment]
Ok, then the workaround above should work for you. The variables need to be in `preview-head` because your `config.js` is executed in the context of preview iframe

**Comment by ndelangen** [comment]
I'm not tied to polished personally, if we can find a way to make things work without polished that works with css custom properties, that'd be a win for all i think.

@Armanio @bennypowers do you know of such a library?

**Comment by bennypowers** [comment]
looks like emotion does

**Comment by domyen** [comment]
We use polished to transform user-supplied colors for use in the theme.  I'd love to support css-vars as well if we can. I'd be open to jettisoning polished if there was a suitable color transformation alternative _or_ if someone can recommend another method. 

As far as I can tell emotion does not do color transformation.

**Comment by ndelangen** [comment]
@bennypowers can you link to docs for that?

PS. can I ask what exactly the point is? I'm not asking in a bad way, I'd like to understand what improvement you're trying to get by moving JS variables into CSS?

You're defining a global scope CSS variable and then referencing that global variable by name in a JS file. I'm sceptical about that being a good idea.

> "Our CSS-in-JS library doesn't support browser standards"
I'm all in favor of using standards, but JS variables are a standard as well?

AFAIK there are no easy ways to read the value of a CSS custom property from JS. If that's possible, we can likely load the value, and keep using polished.

Of course CSS custom properties are designed to be mutable & cascading at runtime.

**Comment by bennypowers** [comment]
> AFAIK there are no easy ways to read the value of a CSS custom property from JS. If that's possible, we can likely load the value, and keep using polished.

Well, you could

```js
const getCssVarVal = cssVar =>
  getComputedStyle(window)
    .getPropertyValue(cssVar)
```

That's all built in to the platform, so it seems pretty easy to me, don't think you need any libraries for that.

> PS. can I ask what exactly the point is? I'm not asking in a bad way, I'd like to understand what improvement you're trying to get by moving JS variables into CSS?

CSS custom properties are an inherent part of CSS the web platform, so it makes sense that a web-development tool that understands them. Your question is sort of like "switch statements can be done as ternary expressions, so we don't support switch statements".

So, from first principles, CSS custom properties should be supported because they are a part of the platform, just like `<table>`, `IntersectionObserver`, or `display: grid;`.

But if you really need the kind of corporate business use case, here's one example of many:

Your company's design department defines their style guide in CSS files, using CSS custom properties. You, as the developer, want to use your company's branding both in your components, but also to brand the storybook interface, so you'd like to load up the theme.css file and have it work properly, just like any other web page.

> You're defining a global scope CSS variable and then referencing that global variable by name in a JS file. I'm sceptical about that being a good idea.

In my case, it's just a CSS value passed as a string, like any other CSS value.

Please forgive errors and typos, as I don't have access to my regular machines.

**Comment by ndelangen** [comment]
I ask because if our code is to become significantly more complex to handle CSS custom properties, we better have a good understanding of the reason. Principle aren't enough reason.

What I'm trying to avoid is that we add complex code that reaches into the runtime global document namespace, does global (possible ui-thread-blocking) operations to convert 1 value to another. And the storybook maintainer team has to maintain that code forever.

I'm worried that the user would be under the assumption they could change the CSS custom property's value at runtime, and those changed would be reflected in storybook's theme. (which would not be the case, we'd read the value just once).

I hope you can understand my concerns.

I interpret your use-case of a design system using CSS custom properties extensively and having some sort of `vars.css` that would be injected into storybook, and thus easily referenced instead of being copied over. I find this a good enough reason to support this, though I'd prefer us to not be responsible for the code that does the transformation.

Would you be interested in providing us with a package/library that turns `'var(--black, black)'` into `'#000'`? We'd be happy to use that, provided it only does the work it needs to do if necessary.

Thank for your time & passion 🙇 

---

PS. storybook runs in more platforms then the browser, it also runs in Node & React-Native.
The theming functionality is shared code.

**Comment by Hypnosphi** [comment]
Which exactly kinds of color transformations do we use? Is there a chance that those usecases can be covered by `opacity`?

**Comment by ndelangen** [comment]
Valid question @Hypnosphi, though I think the options of using the `opacity` property is slim. The way opacity work is it affects the entire element, making it so it only affects the textColor or borderColor means restructuring the html in weird ways.

transforming transparency is best done using `rgba` or `hsla`, `hsla` is also especially useful for transforming hue saturation & lightness. That is what what `hsl` stands for.

transforming rgb values to hsl is possible in JS, afaik not in CSS.

**Comment by Hypnosphi** [comment]
Well you can always put the thing that you want to make half-transparent to a separate element or pseudoelement

**Comment by ndelangen** [comment]
I've just gone through this document, and AFAICS there are no...
`color: lighten(0.2, var(--my-var))`,
`color: darken(0.2, var(--my-var))`,
`color: transparentize(0.2, var(--my-var))`,

https://www.w3.org/TR/css-color-4/#changes-from-3

...in the CSS spec. There are 12 ways of **defining** a color, but no ways of converting 1 to another or slightly adjusting a color after it's been defined.

This seems to be a staple of why css processors are so useful. That, and JS of course.

All code-examples in the document from W3C about the CSS spec, are in JS.
https://www.w3.org/TR/css-color-4/#color-conversion-code

@Hypnosphi Those 3 listed above are what we use.
```js
import { darken, lighten, transparentize } from 'polished';
```

We also use `rgba` but doesn't really do anything other than `(r,g,b,a) => `rgba(${r}, ${g}, ${b}, ${a})`.
We might as well just write that, no transformation, just convenience.

**Comment by Hypnosphi** [comment]
That's why you can't find it in spec
https://github.com/postcss/postcss-color-function#deprecated

And that's how you do darken/lighten using the plugin which still works despite the deprecation
https://github.com/postcss/postcss-color-function#deprecated

**Comment by Hypnosphi** [comment]
> Those 3 listed above are what we use.

Ok, so we're actually only transforming the lightness. This should be pretty easy to achieve using [CSS filters](https://developer.mozilla.org/ru/docs/Web/CSS/filter). The downside is the same as with opacity: you need a separate (pseudo)element for each color usage that you want to transform

**Comment by Hypnosphi** [comment]
> `(r,g,b,a) => `rgba(${r}, ${g}, ${b}, ${a})`;`

This should work OK with custom properties (aka CSS vars)

**Comment by ndelangen** [comment]
SO we're left with 2 scenarios for a potential fix:

**1**:
Never try to read the css value, never try to transform it, just pass it around as a value, unchangeable. In order to do lightness and darkness and hue changes we use pseudo-elements + css filters.

possible side-effects and dangers:
- possibly need to position the pseudo elements in 'funny' ways, to emulate things like borders, underlines, box-shadows, etc.
- filters makes the element have their own rendering stack, which could have side-effects

**2**:
We detect the use of `vars()` in the value and try to resolve the actual color before putting it into the storybook theme.

possible side-effects and dangers:
- Users might do runtime changes to the css vars's value and expect them to have effect. Which they will not.

---

Personally my vote is to NOT do option 1. I find it too fragile, too hacky for it too work for too long. Filters are also CPU & memory hogs, and sometimes cause odd buggy behavior in browser in my experience.

**Comment by Hypnosphi** [comment]
> Users might do runtime changes to the css vars's value and expect them to have effect. Which they will not.

Looks like a proper place for a warning

**Comment by bennypowers** [comment]
FWIW I'm comfortable with a fix that just reads the initial value once. Just please make sure user CSS is loaded before reading the values, and documentation should be super clear on why Storybook doesn't yet fully support this built-in feature of the web platform.

A later, more comprehensive fix might prefer built-in CSS filters to tooling, as @Hypnosphi suggested.

Thanks for your consideration

**Comment by ndelangen** [comment]
Cool, so we come to the consensus that the solution is:

Reading the value before creating the theme just once and warning the user somehow that this is a one-off operation.

@bennypowers my question still stands, would you be able to provide us with the code to do it? Preferable as an external module. Or point us to an existing solution of course.

**Comment by Hypnosphi** [comment]
> Storybook doesn't yet fully support this built-in feature of the web platform

That's just not true. All the web platform fully works in your stories (just because they run in browser). Theming the Storybook UI itself is an additional feature which has its intrinsic limitations. For example, you can't pass gradients as backgrounds. And gradients have been a part of web standarts for quite a while.

**Comment by bennypowers** [comment]
> Storybook UI [...] has its intrinsic limitations

Right, so we agree that this CSS feature isn't supported. No big deal, I just think that should be explicit so developers are informed.

@ndelangen You need something fancier than this?
```js
const getCssVarVal = cssVar =>
  getComputedStyle(document.body)
    .getPropertyValue(cssVar)
```

This will return `black` for, for example 
```css
html {
  --foo: black;
}
```

**Comment by Hypnosphi** [comment]
> Right, so [...] I [..] should be [...] fancier

Ok, I'm able to strip the context as well

**Comment by bennypowers** [comment]
I'm going to bow out of this thread, since it seems to me that the discussion isn't progressing in a particularly productive direction. I really hope this fix gets in soon. Thanks all for your consideration.

**Comment by ndelangen** [comment]
@bennypowers Code looks good to me, if that works for you, my suggestion is to use that code, when you create the theme. Putting that code into the storybook core, feels like unwise. I'll try to explain:

The solution makes a lot of assumptions and have a lot of nuanced limitations.
It assumes:
- the value passed is always a `var(--something)`
- the var exists on body/html element
- there will be a document to read this value from
- users understand the limitation of it being read only once, and being 'bound', and even if we put a warning, that will be noise to many users who do know. Making them pay less attention to warnings that do matter.

Is this an option for you:

```js
import { create } from '@storybook/theming/create';

export default create({
  base: 'light',
  colorSecondary: getCssVarVal('var(--my-var, red)'),
});
```

This works best for everyone right? You get the support, we don't have to worry about this global render-blocking function.

I'm sorry if you feel the thread was/is unproductive.

**Comment by ndelangen** [comment]
This is fixed in 5.2!

**Comment by ndelangen** [comment]
It should be fixed by:
https://github.com/storybookjs/storybook/pull/6939/files

**Comment by carriemorrison** [comment]
Is this still the way to get theming working with css custom properties?

```
import { create } from '@storybook/theming/create';

export default create({
  base: 'light',
  colorSecondary: getCssVarVal('var(--my-var'),
});
```

I'm using `"@storybook/react": "^6.5.14",` and still see issues on some storybook pages eg mdx files...

```
index-681e4b07.js:41 Uncaught Error: Couldn't parse the color string. Please provide the color as a string in hex, rgb, rgba, hsl or hsla notation.


    at parseToRgb (index-681e4b07.js:41:1)
    at parseToHsl (index-681e4b07.js:56:1)
    at darken (index-681e4b07.js:219:1)
    at fn (index-681e4b07.js:195:1)
    at _storybook_theming__WEBPACK_IMPORTED_MODULE_55__.styled.table.isLoading (index-681e4b07.js:1965:1)
    at handleInterpolation (index.js:3482:1)
    at serializeStyles (index.js:3597:1)
    at index.js:3805:1
    at index.js:2509:1
    at renderWithHooks (react-dom.development.js:16305:1)
```
